pkg/flow/internal/controller/loader.go

```diff
@@ -564,84 +568,106 @@ func (l *Loader) OriginalGraph() *dag.Graph {
+	if len(updatedNodes) == 0 {
+		return
+	}
 	tracer := l.tracer.Tracer("")
-
-	l.mut.RLock()
-	defer l.mut.RUnlock()
+	spanCtx, span := tracer.Start(context.Background(), "SubmitDependantsForEvaluation", trace.WithSpanKind(trace.SpanKindInternal))
+	span.SetAttributes(attribute.Int("originators_count", len(updatedNodes)))
+	defer span.End()
 
 	l.cm.controllerEvaluation.Set(1)
 	defer l.cm.controllerEvaluation.Set(0)
-	start := time.Now()
-
-	spanCtx, span := tracer.Start(context.Background(), "GraphEvaluatePartial", trace.WithSpanKind(trace.SpanKindInternal))
-	span.SetAttributes(attribute.String("initiator", c.NodeID()))
-	defer span.End()
 
-	logger := log.With(l.log, "trace_id", span.SpanContext().TraceID())
-	level.Info(logger).Log("msg", "starting partial graph evaluation")
-	defer func() {
-		span.SetStatus(codes.Ok, "")
+	l.mut.RLock()
+	defer l.mut.RUnlock()
 
-		duration := time.Since(start)
-		level.Info(logger).Log("msg", "finished partial graph evaluation", "duration", duration)
-		l.cm.componentEvaluationTime.Observe(duration.Seconds())
-	}()
+	dependenciesToParentsMap := make(map[dag.Node]*ComponentNode)
+	for _, parent := range updatedNodes {
+		// Make sure we're in-sync with the current exports of parent.
+		l.cache.CacheExports(parent.ID(), parent.Exports())
+		// We collect all nodes directly incoming to parent.
+		_ = dag.WalkIncomingNodes(l.graph, parent, func(n dag.Node) error {
+			dependenciesToParentsMap[n] = parent
+			return nil
+		})
+	}
 
-	// Make sure we're in-sync with the current exports of c.
-	l.cache.CacheExports(c.ID(), c.Exports())
+	// Submit all dependencies for asynchronous evaluation.
+	// During evaluation, if a node's exports change, Flow will add it to updated nodes queue (controller.Queue) and
+	// the Flow controller will call EvaluateDependencies on it again. This results in a concurrent breadth-first
+	// traversal of the nodes that need to be evaluated.
+	for n, parent := range dependenciesToParentsMap {
+		dependantCtx, span := tracer.Start(spanCtx, "SubmitForEvaluation", trace.WithSpanKind(trace.SpanKindInternal))
+		span.SetAttributes(attribute.String("node_id", n.NodeID()))
+		span.SetAttributes(attribute.String("originator_id", parent.NodeID()))
 
-	_ = dag.WalkReverse(l.graph, c, func(n dag.Node, parent dag.Node) error {
-		if n == c {
-			// Skip over the starting component; the starting component passed to
-			// EvaluateDependencies had its exports changed and none of its input
-			// arguments will need re-evaluation.
-			return nil
+		// Submit the node for asynchronous evaluation.
+		err := l.workerPool.SubmitWithKey(n.NodeID(), l.concurrentEvalFn(n, dependantCtx, tracer, parent))
+		if err != nil {
+			// The error typically means that the workerPool queue is full. This could mean we have too many components
+			// and the agent cannot keep up with the evaluation. To degrade gracefully, we log the error and continue.
+			// Some evaluations may be missed.
```

Comment:
> Some evaluations may be missed.

Is this a problem? Do we want to make sure we update the whole graph occasionally, on a timed tick, just to guarantee that it's updated in full at least once in a while? E.g. is it possible to have a situation where we miss an evaluation, and then the next evaluation happens after a very long time (because no updates ever came on upstream components)?

Reply:
I wouldn't expect the queue to raise to this level normally and I did consider this as a failure mode... however, having thought more about this, I came up with a way to easily retry the evaluation of components that failed to enqueue and I think it's safe to add it. I wrote a test for it as well and you can check it out in this commit: https://github.com/grafana/agent/pull/5280/commits/5d0862d6e94a6914743f3d7873141e0af1dbf2c6

```
ts=2023-10-03T16:25:15.77487Z level=error msg="failed to submit node for evaluation - the agent is likely overloaded and cannot keep up with evaluating components" controller_id="" err="worker queue is full" node_id=testcomponents.passthrough.inc_dep_2 originator_id=testcomponents.count.inc
```

Thanks for questioning this approach :) My only fear is a retry storm which can lead to some kind of deadlock scenario, but that should be pretty obvious from logs.